File: web/src/types/agents.ts
export type AgentRole = "owner" | "editor" | "viewer";
export type AgentStatus = "draft" | "active" | "archived" | "deleted";
export type AgentChannel = "hosted" | "embed" | "api";
export type AgentAuthMode = "private" | "tenant" | "public" | "token";
export type AgentReleaseStatus =
  | "queued"
  | "running"
  | "passed"
  | "failed"
  | "cancelled"
  | "stale";
export type AgentCapabilityType =
  | "native"
  | "model_native"
  | "mcp"
  | "skill"
  | "connector";

export interface AgentIdentitySpec {
  icon_url?: string | null;
  theme_color?: string | null;
  welcome_message: string;
  suggested_prompts: string[];
}

export interface AgentModelSpec {
  model_id: string;
  provider_id?: string | null;
  temperature?: number | null;
  max_tokens?: number | null;
  thinking_mode?: string | null;
}

export interface AgentCapabilityBinding {
  type: AgentCapabilityType;
  resource_id: string;
  resource_version?: string | null;
  schema_hash?: string | null;
  config: Record<string, unknown>;
}

export interface AgentKnowledgeBinding {
  dataset_id: string;
  retrieval_config: {
    mode?: "auto" | "tool" | "off";
    top_k?: number;
    threshold?: number;
    include_images?: boolean;
  };
}

export interface AgentSpec {
  schema_version: "agent-spec/v1";
  identity: AgentIdentitySpec;
  instructions: string;
  model: AgentModelSpec;
  capabilities: AgentCapabilityBinding[];
  knowledge: AgentKnowledgeBinding[];
  memory: Record<string, unknown>;
}

export interface AgentSummary {
  tenant_id: string;
  agent_id: string;
  slug: string;
  name: string;
  description: string;
  owner_id: string;
  status: AgentStatus;
  caller_role: AgentRole;
  draft_revision: number | null;
  created_at: string;
  updated_at: string;
}

export interface AgentDetail extends AgentSummary {
  draft?: {
    revision: number;
    schema_version: string;
    spec_hash: string;
    updated_at: string;
  } | null;
  archived_at?: string | null;
  deleted_at?: string | null;
}

export interface AgentDraft {
  tenant_id: string;
  draft_id: string;
  agent_id: string;
  revision: number;
  schema_version: string;
  spec: AgentSpec;
  spec_hash: string;
  updated_by: string;
  created_at: string;
  updated_at: string;
}

export interface AgentVersion {
  tenant_id: string;
  agent_version_id: string;
  agent_id: string;
  version_number: number;
  schema_version: string;
  spec: AgentSpec;
  spec_hash: string;
  source_draft_id: string;
  source_draft_revision: number;
  release_evaluation_id?: string | null;
  release_identity_hash?: string | null;
  created_by: string;
  created_at: string;
}

export interface AgentChannelPolicy {
  attachments: boolean;
  high_risk_tools: boolean;
  allowed_origins: string[];
  requests_per_minute?: number;
  requests_per_day?: number;
  ip_requests_per_minute?: number;
  ip_requests_per_day?: number;
  publication_requests_per_minute?: number;
  publication_requests_per_day?: number;
}

export interface AgentApiToken {
  token_id: string;
  publication_id: string;
  name: string;
  scopes: string[];
  expires_at: string | null;
  revoked_at: string | null;
  last_used_at: string | null;
  rotated_from_token_id: string | null;
  created_by: string;
  created_at: string;
}

export interface AgentApiTokenIssue {
  token: string;
  token_metadata: AgentApiToken;
  request_id: string;
}

export interface AgentReleaseFinding {
  code: string;
  field: string;
  message: string;
}

export interface AgentReleaseEvaluationEvent {
  event_id: string;
  evaluation_id: string;
  sequence: number;
  status: AgentReleaseStatus;
  summary: Record<string, unknown>;
  created_at: string;
}

export interface AgentReleaseGate {
  schema_version?: string;
  status: AgentReleaseStatus;
  profile_id?: string;
  profile_version?: string;
  execution_scope?: string;
  model_quality_evaluated?: boolean;
  blocking_findings: AgentReleaseFinding[];
  non_blocking_findings: AgentReleaseFinding[];
  metrics?: {
    critical_pass_rate?: number;
    configured_critical_pass_rate?: number;
    validation_duration_ms?: number;
    provider_cost_cents?: number;
    evaluator_results?: Array<{
      evaluator: string;
      blocking: boolean;
      status: "passed" | "failed";
    }>;
  };
}

export interface AgentReleaseEvaluation {
  tenant_id: string;
  evaluation_id: string;
  agent_id: string;
  draft_id: string;
  draft_revision: number;
  spec_hash: string;
  runtime_fingerprint: Record<string, string>;
  runtime_fingerprint_hash: string;
  release_identity_hash: string;
  evaluation_identity_hash?: string | null;
  profile_id: string;
  profile_version: string;
  dataset_id: string | null;
  dataset_version?: string | null;
  dataset_manifest_hash?: string | null;
  experiment_run_id: string | null;
  channel: AgentChannel;
  auth_mode: AgentAuthMode;
  channel_policy: AgentChannelPolicy;
  channel_policy_hash: string;
  status: AgentReleaseStatus;
  stale: boolean;
  stale_reasons: string[];
  validation_snapshot: Record<string, unknown>;
  gate_snapshot: AgentReleaseGate;
  events: AgentReleaseEvaluationEvent[];
  created_by: string;
  created_at: string;
  started_at?: string | null;
  completed_at?: string | null;
}

export interface AgentEvalDataset {
  dataset_id: string;
  tenant_id: string;
  name: string;
  description: string;
  version: string;
  schema: Record<string, unknown>;
  metadata: Record<string, unknown>;
  updated_at?: string | null;
}

export interface AgentReleaseDiffSection {
  changed: boolean;
  before_hash: string;
  after_hash: string;
  changed_paths: string[];
  before_length?: number;
  after_length?: number;
  before?: unknown;
  after?: unknown;
}

export interface AgentReleaseDiff {
  evaluation_id: string;
  draft_revision: number;
  publication_id: string | null;
  current_version_id: string | null;
  current_version_number: number | null;
  diff: {
    schema_version: string;
    changed_sections: string[];
    sections: Record<string, AgentReleaseDiffSection>;
  };
}

export interface AgentPublication {
  tenant_id: string;
  publication_id: string;
  agent_id: string;
  channel: AgentChannel;
  public_id: string;
  version_id: string | null;
  version_number?: number | null;
  version_spec_hash?: string | null;
  auth_mode: AgentAuthMode;
  policy: AgentChannelPolicy;
  status: "draft" | "active" | "disabled" | "degraded";
  created_by: string;
  updated_by: string;
  created_at: string;
  updated_at: string;
}

export interface AgentPublishEvent {
  event_id: string;
  publication_id: string;
  agent_id: string;
  from_version_id: string | null;
  to_version_id: string;
  actor_id: string;
  reason: string;
  validation_snapshot: Record<string, unknown>;
  operation: "promote" | "rollback";
  release_evaluation_id: string | null;
  request_hash: string | null;
  created_at: string;
}

export interface AgentReleaseMutation {
  request_id: string;
  version: AgentVersion;
  publication: AgentPublication;
  event: AgentPublishEvent;
  idempotent_replay: boolean;
}

export interface AgentPage {
  items: AgentSummary[];
  next_cursor: string | null;
}

export interface AgentRuntimeSession {
  session_id: string;
  agent_id: string;
  agent_version_id: string | null;
  draft_revision: number | null;
  publication_id: string | null;
  channel: "preview" | "hosted" | "embed" | "api";
  runtime_fingerprint: string;
  request_id: string;
}

export interface AgentStreamEvent {
  event?: string;
  event_type?: string;
  content?: string;
  message?: string;
  tool_name?: string;
  status?: string;
  duration_ms?: number;
  dataset_id?: string;
  dataset_name?: string;
  citation_count?: number;
  data?: Record<string, unknown> | string;
  [key: string]: unknown;
}

export interface AgentCatalogTool {
  id: string;
  name: string;
  description: string;
  category: string;
  risk: string;
}

export interface AgentCatalogMcpTool {
  tool_id: string;
  server_id: string;
  server_name: string;
  runtime_name: string;
  description: string;
  snapshot_id: string;
  schema_hash: string;
  risk_level: "low" | "medium" | "high" | "critical";
  connection_id: string | null;
  principal_type: "service_account" | "user_delegated" | null;
  enabled: boolean;
}

export interface AgentCatalogSkill {
  name: string;
  title?: string;
  description?: string;
  version?: string;
  version_id?: string;
  content_hash?: string;
  enabled?: boolean;
  status?: string;
  source?: string;
}

export interface AgentCatalogConnector {
  provider: string;
  display_name: string;
  description?: string | null;
  connected: boolean;
  grant_id?: string;
  principal_type?: "service_account" | "user_delegated";
  scopes?: string[];
  allowed_channels?: string[];
}

export interface AgentOperationsTrace {
  trace_id: string;
  agent_id: string;
  agent_version_id: string | null;
  publication_id: string | null;
  channel: AgentChannel | "preview" | "builtin";
  session_id: string | null;
  status: string;
  model_id: string | null;
  total_latency_ms: number;
  total_tokens: number;
  total_cost_cents: number;
  input_preview: string;
  output_preview: string;
  redaction_state: Record<string, unknown>;
  started_at: string | null;
  created_at: string | null;
}

export interface AgentOperationsMetrics {
  total_runs?: number;
  succeeded_runs?: number;
  failed_runs?: number;
  sessions?: number;
  success_rate?: number | null;
  avg_latency_ms?: number;
  p50_latency_ms?: number;
  p95_latency_ms?: number;
  avg_ttft_ms?: number;
  p50_ttft_ms?: number;
  p95_ttft_ms?: number;
  total_tokens?: number;
  total_cost_cents?: number;
  tool_calls?: number;
  tool_succeeded?: number;
  tool_success_rate?: number | null;
  knowledge_queries?: number;
  knowledge_hits?: number;
  knowledge_hit_rate?: number | null;
  feedback_count?: number;
  positive_feedback_count?: number;
  feedback_positive_rate?: number | null;
  oldest_trace_at?: string | null;
  newest_trace_at?: string | null;
  retention_limited?: boolean;
  retention?: {
    trace_retention_days?: number;
    legal_hold?: boolean;
    last_retention_cleanup_at?: string | null;
  };
  breakdown?: Array<Record<string, unknown>>;
}

export interface AgentAnalyticsResponse {
  agent_id: string;
  caller_role: AgentRole;
  metrics: AgentOperationsMetrics;
  traces: AgentOperationsTrace[];
  total: number;
  limit: number;
  offset: number;
  filters: Record<string, unknown>;
}

export interface AgentAuditEvent {
  id: number;
  user_id: string | null;
  action: string;
  status: string;
  agent_id: string;
  agent_version_id: string | null;
  publication_id: string | null;
  channel: string | null;
  request_summary: Record<string, unknown>;
  response_summary: Record<string, unknown>;
  redaction_state: Record<string, unknown>;
  created_at: string;
}

export interface AgentAuditPage {
  events: AgentAuditEvent[];
  total: number;
  limit: number;
  offset: number;
}

export interface AgentGovernancePolicy {
  tenant_id: string;
  agent_id: string;
  trace_retention_days: number;
  runtime_retention_days: number;
  attachment_retention_days: number;
  legal_hold: boolean;
  principal_requests_per_minute: number;
  principal_requests_per_day: number;
  ip_requests_per_minute: number;
  ip_requests_per_day: number;
  publication_requests_per_minute: number;
  publication_requests_per_day: number;
  max_agents_per_tenant: number;
  max_active_publications: number;
  max_concurrent_runs: number;
  max_daily_tokens: number;
  max_daily_mcp_calls: number;
  max_storage_bytes: number;
  alert_threshold_percent: number;
  cache_epoch: number;
  updated_by: string;
  created_at: string | null;
  updated_at: string | null;
}

export type AgentGovernancePolicyUpdate = Partial<
  Pick<
    AgentGovernancePolicy,
    | "trace_retention_days"
    | "runtime_retention_days"
    | "attachment_retention_days"
    | "legal_hold"
    | "principal_requests_per_minute"
    | "principal_requests_per_day"
    | "ip_requests_per_minute"
    | "ip_requests_per_day"
    | "publication_requests_per_minute"
    | "publication_requests_per_day"
    | "max_agents_per_tenant"
    | "max_active_publications"
    | "max_concurrent_runs"
    | "max_daily_tokens"
    | "max_daily_mcp_calls"
    | "max_storage_bytes"
    | "alert_threshold_percent"
  >
>;

export interface AgentDataDeletion {
  deletion_id: string;
  tenant_id: string;
  agent_id: string;
  scope: "retention" | "user" | "tenant";
  subject_user_id: string | null;
  status: "pending" | "completed" | "failed" | "blocked";
  deleted_counts: Record<string, number>;
  error_code: string | null;
  requested_by: string;
  requested_at: string;
  attempt_count: number;
  last_attempt_at: string | null;
  completed_at: string | null;
}

export interface AgentApiErrorDetail {
  code?: string;
  message?: string;
  request_id?: string;
  current_revision?: number;
  errors?: Array<{ field: string; code: string; message: string }>;
  findings?: AgentReleaseFinding[];
}

export const DEFAULT_AGENT_INSTRUCTIONS =
  "Fulfill the Agent purpose described in its name and description. Follow the configured capability, knowledge, memory, and safety policies.";

export function createDefaultAgentSpec(): AgentSpec {
  return {
    schema_version: "agent-spec/v1",
    identity: {
      icon_url: null,
      theme_color: "#7B7BE8",
      welcome_message: "",
      suggested_prompts: [],
    },
    instructions: DEFAULT_AGENT_INSTRUCTIONS,
    model: {
      model_id: "qwen3.7-plus",
      provider_id: "dashscope",
      temperature: 0.3,
      max_tokens: 4096,
    },
    capabilities: [],
    knowledge: [],
    memory: { mode: "session" },
  };
}

File: src/api/schemas/agents.py
"""Typed public API contracts for Agent Studio lifecycle management."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal
from urllib.parse import urlsplit

from ai_gateway_core.persistence.repositories.agent_repository import (
    redact_agent_spec_for_read,
    unsafe_agent_spec_paths,
)
from ai_gateway_core.security.redaction import redact_trace_text
from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

AgentRole = Literal["owner", "editor", "viewer"]
AgentStatus = Literal["draft", "active", "archived", "deleted"]
CapabilityType = Literal["native", "model_native", "mcp", "skill", "connector", "knowledge"]
AgentChannel = Literal["hosted", "embed", "api"]
AgentAuthMode = Literal["private", "tenant", "public", "token"]
AgentReleaseStatus = Literal["queued", "running", "passed", "failed", "cancelled", "stale"]


class AgentIdentitySpec(BaseModel):
    model_config = ConfigDict(extra="forbid")

    icon_url: str | None = Field(None, max_length=2048)
    theme_color: str | None = Field(None, pattern=r"^#[0-9A-Fa-f]{6}$")
    welcome_message: str = Field("", max_length=4000)
    suggested_prompts: list[str] = Field(default_factory=list, max_length=12)

    @field_validator("suggested_prompts")
    @classmethod
    def _validate_suggested_prompts(cls, value: list[str]) -> list[str]:
        if any(not prompt.strip() or len(prompt) > 500 for prompt in value):
            raise ValueError("suggested prompts must be 1..500 characters")
        return value


class AgentModelSpec(BaseModel):
    model_config = ConfigDict(extra="forbid")

    model_id: str = Field("", max_length=255)
    provider_id: str | None = Field(None, max_length=128)
    temperature: float | None = Field(None, ge=0, le=2)
    max_tokens: int | None = Field(None, ge=1, le=1_000_000)
    thinking_mode: str | None = Field(None, max_length=64)


class AgentCapabilityBinding(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: CapabilityType
    resource_id: str = Field("", max_length=255)
    resource_version: str | None = Field(None, max_length=255)
    schema_hash: str | None = Field(None, pattern=r"^[0-9a-f]{64}$")
    config: dict[str, Any] = Field(default_factory=dict)


class AgentKnowledgeBinding(BaseModel):
    model_config = ConfigDict(extra="forbid")

    dataset_id: str = Field(..., min_length=1, max_length=255)
    retrieval_config: dict[str, Any] = Field(default_factory=dict)


class AgentSpec(BaseModel):
    """Versioned editing shape; runtime resolution is owned by later Phases."""

    model_config = ConfigDict(extra="forbid")

    schema_version: Literal["agent-spec/v1"] = "agent-spec/v1"
    identity: AgentIdentitySpec = Field(default_factory=AgentIdentitySpec)
    instructions: str = Field("", max_length=100_000)
    model: AgentModelSpec = Field(default_factory=AgentModelSpec)
    capabilities: list[AgentCapabilityBinding] = Field(default_factory=list, max_length=256)
    knowledge: list[AgentKnowledgeBinding] = Field(default_factory=list, max_length=128)
    memory: dict[str, Any] = Field(default_factory=dict)

    @model_validator(mode="after")
    def _no_sensitive_values(self) -> AgentSpec:
        unsafe = unsafe_agent_spec_paths(self.model_dump(mode="python"))
        if unsafe:
            raise ValueError(f"{unsafe[0]} must not contain credentials or secrets")
        return self


class AgentCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(..., min_length=1, max_length=255)
    slug: str | None = Field(None, min_length=1, max_length=128)
    description: str = Field("", max_length=4000)
    spec: AgentSpec = Field(default_factory=AgentSpec)


class AgentUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(None, min_length=1, max_length=255)
    slug: str | None = Field(None, min_length=1, max_length=128)
    description: str | None = Field(None, max_length=4000)

    @model_validator(mode="after")
    def _at_least_one_field(self) -> AgentUpdateRequest:
        if not self.model_fields_set:
            raise ValueError("at least one mutable Agent field is required")
        return self


class AgentDraftUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    spec: AgentSpec
    name: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = Field(None, max_length=4000)


class AgentCopyRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(None, min_length=1, max_length=255)
    slug: str | None = Field(None, min_length=1, max_length=128)


class AgentArchiveRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    disable_publications: bool = False


class AgentChannelPolicy(BaseModel):
    model_config = ConfigDict(extra="forbid")

    attachments: bool = False
    high_risk_tools: bool = False
    allowed_origins: list[str] = Field(default_factory=list, max_length=64)
    requests_per_minute: int = Field(default=30, ge=1, le=10_000)
    requests_per_day: int = Field(default=1000, ge=1, le=10_000_000)
    ip_requests_per_minute: int = Field(default=60, ge=1, le=10_000)
    ip_requests_per_day: int = Field(default=2000, ge=1, le=10_000_000)
    publication_requests_per_minute: int = Field(default=300, ge=1, le=100_000)
    publication_requests_per_day: int = Field(default=10_000, ge=1, le=100_000_000)

    @field_validator("allowed_origins")
    @classmethod
    def _validate_origins(cls, value: list[str]) -> list[str]:
        normalized: list[str] = []
        for raw in value:
            origin = raw.strip().rstrip("/")
            parsed = urlsplit(origin)
            if (
                "*" in origin
                or origin.lower() == "null"
                or parsed.scheme not in {"http", "https"}
                or not parsed.netloc
                or parsed.username
                or parsed.password
                or parsed.path
                or parsed.query
                or parsed.fragment
            ):
                raise ValueError("allowed_origins must contain exact http(s) origins")
            normalized.append(origin)
        return sorted(set(normalized))


class AgentReleaseEvaluationRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    draft_revision: int = Field(..., ge=1)
    dataset_id: uuid.UUID | None = None
    channel: AgentChannel = "hosted"
    auth_mode: AgentAuthMode = "private"
    channel_policy: AgentChannelPolicy = Field(default_factory=AgentChannelPolicy)

    @model_validator(mode="after")
    def _public_channel_is_conservative(self) -> AgentReleaseEvaluationRequest:
        if self.auth_mode == "public" and self.channel_policy.high_risk_tools:
            raise ValueError("public channels cannot enable high-risk tools")
        return self


class AgentPublishRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    evaluation_id: uuid.UUID
    reason: str = Field("", max_length=1000)

    @field_validator("reason")
    @classmethod
    def _reason_contains_no_secret_material(cls, value: str) -> str:
        if redact_trace_text(value) != value:
            raise ValueError("release reason must not contain credentials or secrets")
        return value


class AgentRollbackRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    target_version_id: uuid.UUID
    reason: str = Field(..., min_length=1, max_length=1000)

    @field_validator("reason")
    @classmethod
    def _reason_contains_no_secret_material(cls, value: str) -> str:
        if redact_trace_text(value) != value:
            raise ValueError("rollback reason must not contain credentials or secrets")
        return value


class AgentMemberRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    role: AgentRole


class AgentSummary(BaseModel):
    model_config = ConfigDict(extra="ignore")

    tenant_id: str
    agent_id: str
    slug: str
    name: str
    description: str = ""
    owner_id: str
    status: AgentStatus
    caller_role: AgentRole
    draft_revision: int | None = None
    created_at: datetime | str
    updated_at: datetime | str


class AgentDraftMetadata(BaseModel):
    model_config = ConfigDict(extra="ignore")

    revision: int
    schema_version: str
    spec_hash: str
    updated_at: datetime | str


class AgentDetail(AgentSummary):
    draft: AgentDraftMetadata | None = None
    archived_at: datetime | str | None = None
    deleted_at: datetime | str | None = None


class AgentDraftResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    tenant_id: str
    draft_id: str
    agent_id: str
    revision: int
    schema_version: str
    spec: dict[str, Any]
    spec_hash: str
    updated_by: str
    created_at: datetime | str
    updated_at: datetime | str

    @field_validator("spec", mode="before")
    @classmethod
    def _redact_legacy_spec(cls, value: Any) -> dict[str, Any]:
        return redact_agent_spec_for_read(value if isinstance(value, dict) else {})


class AgentVersionResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    tenant_id: str
    agent_version_id: str
    agent_id: str
    version_number: int
    schema_version: str
    spec: AgentSpec
    spec_hash: str
    source_draft_id: str
    source_draft_revision: int
    release_evaluation_id: str | None = None
    release_identity_hash: str | None = None
    created_by: str
    created_at: datetime | str

    @model_validator(mode="before")
    @classmethod
    def _normalize_resolved_spec(cls, value: Any) -> Any:
        if not isinstance(value, dict):
            return value
        normalized = dict(value)
        raw_spec = normalized.get("spec", normalized.get("resolved_spec"))
        normalized["spec"] = redact_agent_spec_for_read(raw_spec)
        return normalized


class AgentReleaseEvaluationEventResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    event_id: str
    evaluation_id: str
    sequence: int
    status: AgentReleaseStatus
    summary: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime | str


class AgentReleaseEvaluationResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    tenant_id: str
    evaluation_id: str
    agent_id: str
    draft_id: str
    draft_revision: int
    spec_hash: str
    runtime_fingerprint: dict[str, Any]
    runtime_fingerprint_hash: str
    release_identity_hash: str
    evaluation_identity_hash: str | None = None
    profile_id: str
    profile_version: str
    dataset_id: str | None = None
    dataset_version: str | None = None
    dataset_manifest_hash: str | None = None
    experiment_run_id: str | None = None
    channel: AgentChannel
    auth_mode: AgentAuthMode
    channel_policy: AgentChannelPolicy
    channel_policy_hash: str
    status: AgentReleaseStatus
    stale: bool = False
    stale_reasons: list[str] = Field(default_factory=list)
    validation_snapshot: dict[str, Any] = Field(default_factory=dict)
    gate_snapshot: dict[str, Any] = Field(default_factory=dict)
    events: list[AgentReleaseEvaluationEventResponse] = Field(default_factory=list)
    created_by: str
    created_at: datetime | str
    started_at: datetime | str | None = None
    completed_at: datetime | str | None = None


class AgentReleaseEvaluationListResponse(BaseModel):
    evaluations: list[AgentReleaseEvaluationResponse] = Field(default_factory=list)


class AgentReleaseDiffResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    evaluation_id: str
    draft_revision: int
    publication_id: str | None = None
    current_version_id: str | None = None
    current_version_number: int | None = None
    diff: dict[str, Any]


class AgentPublicationResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    tenant_id: str
    publication_id: str
    agent_id: str
    channel: AgentChannel
    public_id: str
    version_id: str | None = None
    version_number: int | None = None
    version_spec_hash: str | None = None
    auth_mode: AgentAuthMode
    policy: dict[str, Any] = Field(default_factory=dict)
    status: Literal["draft", "active", "disabled", "degraded"]
    created_by: str
    updated_by: str
    created_at: datetime | str
    updated_at: datetime | str


class AgentPublishEventResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    event_id: str
    publication_id: str
    agent_id: str
    from_version_id: str | None = None
    to_version_id: str
    actor_id: str
    reason: str = ""
    validation_snapshot: dict[str, Any] = Field(default_factory=dict)
    operation: Literal["promote", "rollback"] = "promote"
    release_evaluation_id: str | None = None
    request_hash: str | None = None
    created_at: datetime | str


class AgentReleaseMutationResponse(BaseModel):
    request_id: str
    version: AgentVersionResponse
    publication: AgentPublicationResponse
    event: AgentPublishEventResponse
    idempotent_replay: bool = False


class AgentMemberResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    tenant_id: str
    agent_id: str
    principal_type: Literal["user", "group"]
    principal_id: str
    role: AgentRole
    created_by: str
    created_at: datetime | str
    updated_at: datetime | str


class AgentPageResponse(BaseModel):
    items: list[AgentSummary]
    next_cursor: str | None = None


class AgentValidationIssue(BaseModel):
    field: str
    code: str
    message: str


class AgentValidationResponse(BaseModel):
    valid: bool
    revision: int
    spec_hash: str
    errors: list[AgentValidationIssue]


class AgentMutationResponse(BaseModel):
    request_id: str
    agent: AgentSummary | AgentDetail


class AgentDraftMutationResponse(BaseModel):
    request_id: str
    draft: AgentDraftResponse


class AgentVersionMutationResponse(BaseModel):
    request_id: str
    version: AgentVersionResponse


class AgentMemberMutationResponse(BaseModel):
    request_id: str
    member: AgentMemberResponse


class AgentStatusResponse(BaseModel):
    request_id: str
    status: str


class AgentGovernancePolicyUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    trace_retention_days: int | None = Field(None, ge=1, le=3650)
    runtime_retention_days: int | None = Field(None, ge=1, le=3650)
    attachment_retention_days: int | None = Field(None, ge=1, le=3650)
    legal_hold: bool | None = None
    principal_requests_per_minute: int | None = Field(None, ge=1, le=1_000_000)
    principal_requests_per_day: int | None = Field(None, ge=1, le=100_000_000)
    ip_requests_per_minute: int | None = Field(None, ge=1, le=1_000_000)
    ip_requests_per_day: int | None = Field(None, ge=1, le=100_000_000)
    publication_requests_per_minute: int | None = Field(None, ge=1, le=1_000_000)
    publication_requests_per_day: int | None = Field(None, ge=1, le=100_000_000)
    max_agents_per_tenant: int | None = Field(None, ge=1, le=100_000)
    max_active_publications: int | None = Field(None, ge=1, le=100_000)
    max_concurrent_runs: int | None = Field(None, ge=1, le=1_000_000)
    max_daily_tokens: int | None = Field(None, ge=1, le=10_000_000_000_000)
    max_daily_mcp_calls: int | None = Field(None, ge=1, le=10_000_000_000)
    max_storage_bytes: int | None = Field(None, ge=1, le=10_000_000_000_000)
    alert_threshold_percent: int | None = Field(None, ge=1, le=100)


class AgentGovernancePolicyResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    tenant_id: str
    agent_id: str
    trace_retention_days: int
    runtime_retention_days: int
    attachment_retention_days: int
    legal_hold: bool
    principal_requests_per_minute: int
    principal_requests_per_day: int
    ip_requests_per_minute: int
    ip_requests_per_day: int
    publication_requests_per_minute: int
    publication_requests_per_day: int
    max_agents_per_tenant: int
    max_active_publications: int
    max_concurrent_runs: int
    max_daily_tokens: int
    max_daily_mcp_calls: int
    max_storage_bytes: int
    alert_threshold_percent: int
    cache_epoch: int = 0
    updated_by: str
    created_at: datetime | str | None = None
    updated_at: datetime | str | None = None


class AgentOperationsTrace(BaseModel):
    model_config = ConfigDict(extra="ignore")

    trace_id: str
    agent_id: str
    agent_version_id: str | None = None
    publication_id: str | None = None
    channel: Literal["preview", "hosted", "embed", "api", "builtin"]
    session_id: str | None = None
    status: str
    model_id: str | None = None
    total_latency_ms: int = 0
    total_tokens: int = 0
    total_cost_cents: int = 0
    input_preview: str = ""
    output_preview: str = ""
    redaction_state: dict[str, Any] = Field(default_factory=dict)
    started_at: datetime | str | None = None
    created_at: datetime | str | None = None


class AgentAnalyticsResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    agent_id: str
    caller_role: AgentRole
    metrics: dict[str, Any]
    traces: list[AgentOperationsTrace]
    total: int
    limit: int
    offset: int
    filters: dict[str, Any] = Field(default_factory=dict)


class AgentAuditEventResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: int
    user_id: str | None = None
    action: str
    status: str
    agent_id: str
    agent_version_id: str | None = None
    publication_id: str | None = None
    channel: str | None = None
    request_summary: dict[str, Any] = Field(default_factory=dict)
    response_summary: dict[str, Any] = Field(default_factory=dict)
    redaction_state: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime | str


class AgentAuditPageResponse(BaseModel):
    events: list[AgentAuditEventResponse]
    total: int
    limit: int
    offset: int


class AgentDataDeletionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    scope: Literal["retention", "user", "tenant"]
    subject_user_id: str | None = Field(None, min_length=1, max_length=255)
    idempotency_key: str = Field(..., min_length=8, max_length=255)

    @model_validator(mode="after")
    def _validate_subject(self) -> AgentDataDeletionRequest:
        if (self.scope == "user") != bool(self.subject_user_id):
            raise ValueError("subject_user_id is required only for user scope")
        return self


class AgentDataDeletionResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")

    deletion_id: str
    tenant_id: str
    agent_id: str
    scope: Literal["retention", "user", "tenant"]
    subject_user_id: str | None = None
    status: Literal["pending", "completed", "failed", "blocked"]
    deleted_counts: dict[str, Any] = Field(default_factory=dict)
    error_code: str | None = None
    requested_by: str
    requested_at: datetime | str
    attempt_count: int = 0
    last_attempt_at: datetime | str | None = None
    completed_at: datetime | str | None = None
    retryable: bool = False

    @model_validator(mode="after")
    def _derive_retryable(self) -> AgentDataDeletionResponse:
        self.retryable = self.status == "failed" and self.completed_at is None
        return self


class AgentCredentialRevocationResponse(BaseModel):
    request_id: str
    revoked: dict[str, int]


class AgentCacheInvalidationResponse(BaseModel):
    request_id: str
    cache_epoch: int
    deleted_cache_rows: int


class AgentErrorDetail(BaseModel):
    code: str
    message: str
    request_id: str
    current_revision: int | None = None
    errors: list[AgentValidationIssue] | None = None
    findings: list[dict[str, Any]] | None = None


class AgentErrorResponse(BaseModel):
    detail: AgentErrorDetail


__all__ = [name for name in globals() if name.startswith("Agent")]
